File: app/src/main/java/br/com/jonathanzanella/myexpenses/helpers/CurrencyTextWatch.java
package br.com.jonathanzanella.myexpenses.helpers;

import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;

public class CurrencyTextWatch implements TextWatcher {
	private final EditText edit;
	private String current;

	public CurrencyTextWatch(EditText edit) {
		this.edit = edit;
	}

	@Override
	public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

	@Override
	public void onTextChanged(CharSequence s, int start, int before, int count) {
		if(!s.toString().equals(current)){
			edit.removeTextChangedListener(this);

			String cleanString = s.toString().replaceAll("[^\\d]", "");

			double parsed = Double.parseDouble(cleanString);
			String formatted = CurrencyHelper.format(parsed);

			current = formatted;
			edit.setText(formatted);
			edit.setSelection(formatted.length());

			edit.addTextChangedListener(this);
		}
	}

	@Override
	public void afterTextChanged(Editable s) {}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/helpers/CurrencyTextWatch.kt
package br.com.jonathanzanella.myexpenses.helpers

import android.text.Editable
import android.text.TextWatcher
import android.widget.EditText

class CurrencyTextWatch(private val edit: EditText) : TextWatcher {
    private var current: String? = null

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        if (s.toString() != current) {
            edit.removeTextChangedListener(this)

            val cleanString = s.toString().replace("[^\\d]".toRegex(), "")

            var formatted = ""
            if (!cleanString.isEmpty()) {
                val parsed = java.lang.Double.parseDouble(cleanString)
                formatted = parsed.toInt().toCurrencyFormatted()
            }

            current = formatted
            edit.setText(formatted)
            edit.setSelection(formatted.length)

            edit.addTextChangedListener(this)
        }
    }

    override fun afterTextChanged(s: Editable) {}
}
